### PR TITLE
aya-bpf: introduce argument coercion

### PR DESCRIPTION
**Pull request by willfindlay**
This PR introduces argument coercion for pt_regs and BTF program types and implements it for kprobes, uprobes, LSM programs, and BTF tracepoints.

You can see a [more concrete example of using this API here][1].

Features:
  - Uses a FromBtfArgs trait with blanket implementations to enable coercion to both primitive types _and_ raw pointers
  - No need for the user to perform any kind of manual casting on the BPF side
  - Similar logic for pt_regs programs, abstracting over the PT_REGS access

TODO:
  - Need to support architectures other than x86 for k/uprobe arguments (bindings need to be fixed)
  
While I was at it, I modified the Cargo.toml file of aya-bpf-macros to fix broken doctests. Perhaps we should add doctests to the CI workflow?

Closes #21

[1]: https://github.com/willfindlay/aya-lsm-args-example

### CODE REVIEW

File: bpf/aya-bpf/src/btf_args.rs
Comment (alessandrod) [inline bpf/aya-bpf/src/btf_args.rs:22]: I'm not sure I understand, shouldn't it be `ctx.add((1 + position) * mem::size_of::<*const c_void>()) as _`?

Comment (willfindlay) [reply bpf/aya-bpf/src/btf_args.rs:22]: Good catch, I'll fix it

Comment (willfindlay) [reply bpf/aya-bpf/src/btf_args.rs]: @alessandrod I've updated this to be more semantically meaningful. While I was at it I just switched to use ptr.offset instead of ptr.add. Much simpler and more correct.

**Comment by willfindlay** [comment]
@alessandrod Any thoughts on how this looks from an API perspective? Once I have the green light I will add supports for `pt_regs`-style contexts. Tracepoints are a bit trickier, since their contexts are actually just interpreted as a pointer to a tracepoint-specific C struct, so the API might diverge a bit there.

**Comment by alessandrod** [comment]
It generally looks great! I think you should go ahead and try it with `pt_regs` as that will inform the design more. For example I think you'll need something like `FromBtfArgument` there too, so that might have to be renamed. 

Then I'd probably call it `ctx.arg(n)` instead of argument, for consistency with `std::env::args()`, `.arg()` from clap etc.

**Comment by willfindlay** [comment]
@alessandrod I've added support for pt_regs probes. I have some example code over at https://github.com/willfindlay/aya-kprobe-arg-example demonstrating the API on a try_to_wake_up kprobe.

I'll definitely squash these commits before we land this, but I'm leaving the history mostly as-is for now to make things easier.

File: bpf/aya-bpf/src/args.rs
Comment (alessandrod) [inline bpf/aya-bpf/src/args.rs:24]: Why is `n: isize` and not `usize`? (here and in other places)

File: bpf/aya-bpf/src/args.rs
Comment (alessandrod) [inline bpf/aya-bpf/src/args.rs:111]: Shouldn't this be `if n <= 6`? (r0-r7)

File: bpf/aya-bpf/src/programs/lsm.rs
Comment (alessandrod) [inline bpf/aya-bpf/src/programs/lsm.rs:14]: I think this is a bit too technical, I'd rephrase it as:

```
/// Returns the `n`th argument passed to the LSM hook, starting from 0.
```

File: bpf/aya-bpf/src/programs/probe.rs
Comment (alessandrod) [inline bpf/aya-bpf/src/programs/probe.rs:22]: And this as:

```
Returns the `n`th argument passed to the probed function, starting from 0.
```

File: bpf/aya-bpf/src/programs/tp_btf.rs
Comment (alessandrod) [inline bpf/aya-bpf/src/programs/tp_btf.rs:14]:
```
Returns the `n`th argument passed to the tracepoint, starting from 0.
```

**Review by alessandrod — CHANGES_REQUESTED** [review]

File: bpf/aya-bpf/src/args.rs
Comment (willfindlay) [reply bpf/aya-bpf/src/args.rs:24]: isize is needed for `ptr::offset`

Comment (willfindlay) [reply bpf/aya-bpf/src/args.rs:24]: other places is just for api consistency

File: bpf/aya-bpf/src/programs/lsm.rs
Comment (willfindlay) [reply bpf/aya-bpf/src/programs/lsm.rs:14]: Sure

File: bpf/aya-bpf/src/args.rs
Comment (willfindlay) [reply bpf/aya-bpf/src/args.rs:24]: I suppose we could also accept a usize and just cast to an isize here though, up to you

Comment (alessandrod) [reply bpf/aya-bpf/src/args.rs:24]: That's an implementation detail leaking out tho, we should take `usize` since you can't get negative arguments, and cast as `isize` in the impl. We don't allow large usize values anyway.

File: bpf/aya-bpf/src/args.rs
Comment (willfindlay) [reply bpf/aya-bpf/src/args.rs:111]: libbpf only defines PT_REGSPARM0 up to PT_REGS_PARM5, so I did the same here. See https://github.com/libbpf/libbpf/blob/17d7f04e7cb8cca664dfccd27c3848d105c56b5e/src/bpf_tracing.h#L200-L222. Are these extra registers documented somewhere?

File: bpf/aya-bpf/src/programs/probe.rs
Comment (willfindlay) [reply bpf/aya-bpf/src/programs/probe.rs:22]: Sure

File: bpf/aya-bpf/src/args.rs
Comment (alessandrod) [reply bpf/aya-bpf/src/args.rs:111]: if we go the libbpf route, then it should be `n <= 4`. See https://en.wikipedia.org/wiki/Calling_convention

**Comment by willfindlay** [comment]
@alessandrod I've applied your documentation nits, changed all signatures to accept a `usize` rather than an `isize`, and also changed the maximum allowed `n` for aarch64 and armv7. I mostly followed https://man7.org/linux/man-pages/man2/syscall.2.html (see the section labelled Architecture calling convention).  Hopefully, I haven't gotten anything wrong, since we're kind of going off the reservation w.r.t. libbpf.

P.S. I'm noticing a pattern where PRs I work on after 2am all have off-by-one errors, so maybe it's time to rethink my sleeping habits :)

**Comment by willfindlay** [comment]
Ah clippy doesn't like casting for offsets, I'll refactor.

File: bpf/aya-bpf/src/args.rs
Comment (alessandrod) [inline bpf/aya-bpf/src/args.rs:108]: I think this should be `regs[0]`?

File: bpf/aya-bpf/src/args.rs
Comment (alessandrod) [inline bpf/aya-bpf/src/args.rs:98]: I don't think this link is too relevant - kprobes can execute in non-syscall contexts, and uprobes use `ProbeContext` too

File: bpf/aya-bpf/src/args.rs
Comment (alessandrod) [inline bpf/aya-bpf/src/args.rs:125]: I think this should be `regs[0]` too

File: bpf/aya-bpf/src/programs/probe.rs
Comment (alessandrod) [inline bpf/aya-bpf/src/programs/probe.rs:47]: I'd phrase this as:

```
Returns the return value of the probed function.
```

**Review by alessandrod — CHANGES_REQUESTED** [review]
I'm starting to think that we should probably not try to implement `ProbeContext::ret()` at all. The registers used for return values depend on the return type (eg int vs float), and can actually span multiple registers.

File: bpf/aya-bpf/src/args.rs
Comment (willfindlay) [reply bpf/aya-bpf/src/args.rs:108]: Ah whoops good catch

File: bpf/aya-bpf/src/args.rs
Comment (willfindlay) [reply bpf/aya-bpf/src/args.rs:98]: Right but this at least appears to be the calling convention they are referencing in libbpf so I figured this would be the best place to consult for pt_regs-related accesses beyond the scope of what libbpf does (just the first five arguments). Alternatively, we could just truncate everything to 5 arguments like libbpf does and call it a day.

File: bpf/aya-bpf/src/args.rs
Comment (alessandrod) [inline bpf/aya-bpf/src/args.rs:182]: this should be `regs[0]`

File: bpf/aya-bpf/src/args.rs
Comment (alessandrod) [inline bpf/aya-bpf/src/args.rs:164]: This should be `regs[0]`

**Review by alessandrod — CHANGES_REQUESTED** [review]
Oops we missed a couple. This should be the last one 🤞

**Comment by willfindlay** [comment]
> Oops we missed a couple. This should be the last one 🤞

Ah whoops. I'll have some time to fix this later tonight!

**Comment by willfindlay** [comment]
@alessandrod should be fixed now

**Review by alessandrod — APPROVED** [review]

**Comment by alessandrod** [comment]
Great work! 🙏

**Pull request merged**
